features/123-chat-push-notification/api.md

```diff
@@ -1,158 +1,108 @@
-  - `skygearChat.afterConversationDeleted((conversation, participants) => {})`
-  - `skygearChat.afterUsersAddedToConversation((conversation, participants, users) => {})`
-  - `skygearChat.afterUsersRemovedFromConversation((conversation, participants, users) => {})`
-
-- Conversation
-  - `@chat.after_message_sent`
-  - `@chat.after_message_updated`
-  - `@chat.after_message_deleted`
-  - `@chat.typing_started`
-  - `@chat.after_conversation_created`
-  - `@chat.after_conversation_updated`
-  - `@chat.after_conversation_deleted`
-  - `@chat.after_users_added_to_conversation`
-  - `@chat.after_users_removed_from_conversation`
+## Task
 
-## Implementation Details
+1. Define new payload format
+2. Update skygear server to support new payload.
 
+## New Payload
 
-### Changes on JS SDK
-- Example:
+The following is the proposed payload.
 
-```javascript
-export function afterMessageSent(callback) {
-  registry.registerHandler('chat:after_message_sent', (req) => {
-    var conversation = blahblahblah ;
-    var message = blahblahblah;
-    callback(message, conversation);
-  }, {});
-}
-```
+### Main Payload
 
-### Changes on Plugin (Hook Wrapper)
-- Wrapper of op wrapper
+Main Payload is a collection of payloads.
 
-- Example:
+|key | type |description           |default|
+|----|------|----------------------|-------|
+|platform|dict|See platform payload|null|
+|message|dict|See message payload|null|
+|data|dict|Custom dictionary by developer|null|
+|custom|dict|See custom payload|null|
+|dry\_run|boolean|dry run only if true|false|
 
-```python
-def after_message_sent(*args, **kwargs):
-    def chat_op(func):
-        _registry.register_op('chat:after_message_sent', func, *args, **kwargs)
-        return func
-    return chat_op
-```
 
 
-### Changes on Plugin (Database Operation)
-- Send message to call specific hook after database operation
-- Example:
+### Platform Payload
 
-``` python
-container.send_action('chat:after_conversation_hook', {'roles': message, 'conversation': conversation} )
-```
+|key | type |description|default|
+|----|------|-----------|-------|
+|android|boolean|send to android device if true|true|
+|ios|boolean|send to android device if true|true|
+
+### Message Payload
+
+|key|type|description|default|
+|---|----|-----------|-------|
+|body|string|message body|null|
+|title|string|message title|null|
+|subtitle|string|message subtitle|null|
+|sound|string|message sound|null|
+|title\_loc\_key|string|key to localized string|null|
+|title\_loc\_args|list of string|arguments of string replacement of title\_loc\_key|null|
+|priority|string|priority of message|"normal"|
+|expiry|int|time to live in seconds| 2419200 (4 weeks)|
 
-### Changes on API at skygear-server
 
-Nil
+### Custom Payload
+|key|type|description|default|
+|---|----|-----------|-------|
+|android\_channel\_id|string| |null|
+|android\_click\_action|string| |null|
+|android\_body\_loc\_key\_action|string| |null|
+|android\_body\_loc\_key\_args|string| |null|
```

Comment:
I don't get why loc_key is android specific

Reply:
I think `body_loc_key` only exists in android?

Reply:
Yes. But i think why I don't get it is because `action-loc-key` is iOS specific. but we don't make it `ios-action-loc-key`.

So may be the real question is when to have device prefix, when not.

Reply:
hmmm, i did not include `action-loc-key`, did you mean `title_loc_key`? because I think it can be unified. If not, then I just split them into `android_title_loc_key` and `ios_title_loc_key`. The rule of thumb is, if a field can be unified / across platform, then a prefix skipped, else I would add a platform prefix.